From eef8e97dd335beccb23f3b13d4f4ad715f23f359 Mon Sep 17 00:00:00 2001
From: Yuxuan Shui <yshui@codeweavers.com>
Date: Fri, 20 Mar 2026 09:23:46 +0000
Subject: [PATCH] crypt32: Don't access
 CERT_CHAIN_ENGINE_CONFIG::dwExclusiveFlags without checking size.

Found by ASan.
---
 dlls/crypt32/chain.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/dlls/crypt32/chain.c b/dlls/crypt32/chain.c
index 6e0e55b9e5a4..cb9d0c7acb36 100644
--- a/dlls/crypt32/chain.c
+++ b/dlls/crypt32/chain.c
@@ -236,17 +236,22 @@ BOOL WINAPI CertCreateCertificateChainEngine(PCERT_CHAIN_ENGINE_CONFIG pConfig,
     TRACE("dwUrlRetrievalTimeout %lu\n", pConfig->dwUrlRetrievalTimeout);
     TRACE("MaximumCachedCertificates %lu\n", pConfig->MaximumCachedCertificates);
     TRACE("CycleDetectionModulus %lu\n", pConfig->CycleDetectionModulus);
-    TRACE("hExclusiveRoot %p\n", pConfig->hExclusiveRoot);
-    TRACE("hExclusiveTrustedPeople %p\n", pConfig->hExclusiveTrustedPeople);
-    TRACE("dwExclusiveFlags %lx\n", pConfig->dwExclusiveFlags);
-    if (pConfig->dwExclusiveFlags) FIXME("dwExclusiveFlags %lx not supported\n", pConfig->dwExclusiveFlags);
-
     if (pConfig->cbSize != sizeof(CERT_CHAIN_ENGINE_CONFIG_NO_EXCLUSIVE_ROOT)
      && pConfig->cbSize != sizeof(CERT_CHAIN_ENGINE_CONFIG))
     {
         SetLastError(E_INVALIDARG);
         return FALSE;
     }
+
+    if (pConfig->cbSize == sizeof(CERT_CHAIN_ENGINE_CONFIG))
+    {
+        TRACE("hExclusiveRoot %p\n", pConfig->hExclusiveRoot);
+        TRACE("hExclusiveTrustedPeople %p\n", pConfig->hExclusiveTrustedPeople);
+        TRACE("dwExclusiveFlags %lx\n", pConfig->dwExclusiveFlags);
+        if (pConfig->dwExclusiveFlags)
+            FIXME("dwExclusiveFlags %lx not supported\n", pConfig->dwExclusiveFlags);
+    }
+
     ret = CRYPT_CheckRestrictedRoot(pConfig->hRestrictedRoot);
     if (!ret)
     {
